pkg/natdiscovery/request_handle.go

```diff
@@ -0,0 +1,156 @@
+/*
+© 2021 Red Hat, Inc. and others.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package natdiscovery
+
+import (
+	"net"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	proto2 "google.golang.org/protobuf/proto"
+	"k8s.io/klog"
+
+	"github.com/submariner-io/submariner/pkg/natdiscovery/proto"
+)
+
+func (nd *natDiscovery) handleRequestFromAddress(req *proto.SubmarinerNatDiscoveryRequest, addr *net.UDPAddr) error {
+	response := proto.SubmarinerNatDiscoveryResponse{
+		RequestNumber: req.RequestNumber,
+		Sender: &proto.EndpointDetails{
+			ClusterId:  nd.localEndpoint.Spec.ClusterID,
+			EndpointId: nd.localEndpoint.Spec.CableName,
+		},
+		Receiver: req.Sender,
+		ReceivedSrc: &proto.IPPortPair{
+			Port: uint32(addr.Port),
+			IP:   addr.IP.String(),
+		},
+	}
+
+	if req.Receiver == nil || req.Sender == nil || req.UsingDst == nil || req.UsingSrc == nil {
+		klog.Warningf("received NAT discovery packet %#v from %#v which seems to be malformed ", req, addr)
+
+		response.Response = proto.ResponseType_MALFORMED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if req.Receiver.GetClusterId() != nd.localEndpoint.Spec.ClusterID {
+		klog.Warningf("received NAT discovery packet for cluster %q, but we are cluster %q", req.Receiver.GetClusterId(),
+			nd.localEndpoint.Spec.ClusterID)
+
+		response.Response = proto.ResponseType_UNKNOWN_DST_CLUSTER
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if req.Receiver.GetEndpointId() != nd.localEndpoint.Spec.CableName {
+		klog.Warningf("received NAT discovery packet for endpoint %q, but we are endpoint %q "+
+			"if the port for NAT discovery has been mapped somewhere an error may exist", req.Receiver.GetEndpointId(),
+			nd.localEndpoint.Spec.CableName)
+
+		response.Response = proto.ResponseType_UNKNOWN_DST_ENDPOINT
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if !nd.isKnownCluster(req.Sender.GetClusterId()) {
```

Comment:
Is this necessary since we verify the `EndpointId` below? If so then `isKnownCluster` and `isKnownEndpoint` should be consistent, ie `isKnownCluster` should also lookup the in `remoteEndpoints` by `EndpointId` instead of searching the map. In fact you could just combine the two:

```
ep := nd.getRemoteEndpointNAT(endpointID)
return ep != nil && ep.Endpoint.Spec.ClusterID == clusterID
```